File: snuqs/buffer/memory_buffer.py
from .buffer import Buffer
from typing import Union
import snuqs._C


class MemoryBuffer(Buffer):
    def __init__(self, size: int):
        self.impl = snuqs._C.MemoryBuffer(size)

    def __getitem__(self, key: int):
        if isinstance(key, slice):
            raise NotImplementedError("slice is not supported")
        else:
            return self.impl.__getitem__(key)

    def __setitem__(self, key: Union[int, slice], val: float):
        if isinstance(key, slice):
            raise NotImplementedError("slice is not supported")
        else:
            return self.impl.__setitem__(key, val)

File: src/buffer/memory_buffer.cc
#include "assertion.h"
#include "buffer.h"

#include <cstdlib>
#include <new>

namespace snuqs {

//
// Memory Buffer
//
template <typename T>
MemoryBuffer<T>::MemoryBuffer(size_t count) : count_(count) {
  buf_ = reinterpret_cast<std::complex<T> *>(
      malloc(sizeof(std::complex<T>) * count));
  if (buf_ == nullptr) {
    throw std::bad_alloc();
  }
}

template <typename T> MemoryBuffer<T>::~MemoryBuffer() {
  if (buf_ != nullptr)
    free(buf_);
}

template <typename T> std::complex<T> *MemoryBuffer<T>::ptr() { return buf_; }

template <typename T> std::complex<T> MemoryBuffer<T>::__getitem__(size_t key) {
  return buf_[key];
}

template <typename T>
void MemoryBuffer<T>::__setitem__(size_t key, std::complex<T> val) {
  buf_[key] = val;
}

template class MemoryBuffer<float>;
template class MemoryBuffer<double>;

} // namespace snuqs
